### ISSUE TITLE
Audio-only stream fails to start

### ISSUE THREAD

**Issue by chrisdeely** [jwplayer/jwplayer#679]
I discovered an odd bug while trying to play an audio-only RTMP stream. An audio-only stream will fail to play (at least audibly) until after the library has been used to play at least one video asset/stream.

The audio player will initialize, buffer & "play" but no sound is heard and the controls have no effect. Once you call ```setup()``` on a video instance, the audio will suddenly work.

I'm attempting to track down the issue, my only clue so far is that the ```firstFrame``` event is not fired for the audio player until after the video instance is initialized.

Below is a simple test page I created for this bug. You will need to supply an appropriate path for the JW js file.

```html
<!DOCTYPE html>
<html>
<head lang="en">
    <meta charset="UTF-8">
    <title>JWP Bug Audio stream fails to start</title>
    <script src="//content.jwplatform.com/libraries/INSERT_LIB_LOCATION.js"></script>
</head>
<body>
    <h1>Bug Description</h1>
    <p>An audio-only live stream (via RTMP or HLS) will not begin playback until after the
        JWPlayer has been used to play a video asset/stream first.</p>
    <h2>Steps to recreate:</h2>
    <ol>
        <li>Click the 'Audio' button to initialize the audio-only player</li>
        <li>The log will show events for the audio player up to and including 'play', but no audio is heard</li>
        <li>Click the 'Video' button</li>
        <li>The audio stream will be heard immediately (note the 'firstFrame' event in the log</li>
    </ol>
      <h1>Audio only test</h1>
      <div id="audio-player"></div>

      <h1>Video test</h1>
      <div id="video-player"></div>

    <button id="play-audio">Audio</button>
    <button id="play-video">Video</button>

    <h2>Log Area</h2>
    <textarea id="log" style="width:400px;height:300px"></textarea>

<script>
    var audioStartBtn = document.getElementById("play-audio");
    var videoStartBtn = document.getElementById("play-video");
    var logArea = document.getElementById("log");

    audioStartBtn.addEventListener("click", function () {
        jwplayer("audio-player").setup({
            height: 40,
            width: 600,
            autostart: true,
            mute: false,
            primary:'flash',
            playlist: [
                {
                    title: "Live Audio Test",
                    sources:[
                        {file: "rtmp://stream-7.antenne.de/liveedge/rockantenne", type:'rtmp'}
                    ]
                }
            ],
            rtmp:{
                subscribe: true
            },
            controls:true,
            skin: {"name": "six"}
        }).on('all', function (type, data) {
            //filter out the very frequent events for performance
            if(type == 'time' || type== 'meta') return;
            console.log(type, data);
            logArea.value += "\n" + type;
        });
    });

    videoStartBtn.addEventListener("click", function () {
        jwplayer("video-player").setup({
            width: 640,
            height: 480,
            autostart: false,
            primary: "flash",
            playlist:[
                {
                    sources:[
                        {file: "http://content.jwplatform.com/videos/nPripu9l-twSo9iFz.mp4"}
                    ]
                }
            ]
        });
    });

</script>
</body>
</html>
```

**Comment by chrisdeely** [comment]
I've just attempted to recreate this issue using the latest ```master``` version, and have been unable to do so. In fact, I tried each version back to 7.0.0 and the audio-only player works just as expected.

It seems that the issue must be with the commercial version hosted by JW, so I'll open a ticket with their support.

**Comment by donato** [comment]
Hi @chrisdeely can you confirm with this version of our commercial player?
http://p.jwpcdn.com/player/v/7.0.2/jwplayer.js

What browser are you using?

**Comment by chrisdeely** [comment]
Hi @donato I have tested with the provided version (and my commercial key) and confirmed that the bug is there.  I have been testing in Chrome 45 beta on OSX

**Comment by donato** [comment]
I just tested it myself on Chrome 44 OSX, and the audio is working.

Can you tell me the result of this method? 
```js
jwplayer('audio-player').getProvider()
```

**Comment by chrisdeely** [comment]
Object {name: "flash_rtmp"}

**Comment by donato** [comment]
That's correct, have you tried updating your flash plugin? Or is it
possible your ad blocker is stopping it?

**Comment by donato** [comment]
Ahh good point!

We anticipated the effect of this on banner ads, although I didn't think of
audio only until you mentioned it.

http://www.jwplayer.com/blog/chrome-power-saving/

**Comment by chrisdeely** [comment]
Odd that this does not seem to effect local builds. I wonder if loading the library from a different domain is having an effect. 

I'll test some other configurations & player dimensions to see if I can find the tipping point.

**Comment by chrisdeely** [comment]
Similar to what is described in the article @donato referenced, the minimum player size I've found that circumvents this behavior is 398x298. 

The dimensions are evaluated independently, meaning the overall area is not the deciding factor.

**Comment by chrisdeely** [comment]
I've been able to confirm that self-hosting the ```swf``` for the Flash provider on the same domain as the containing is sufficient to avoid automatic pausing by Chrome.

I am able to load the CDN-hosted js library from the JW Platform, and override the ```swf``` location using the following config. Run this code after the JW library is loaded, but before creating any instances.

```javascript
jwplayer.defaults = {
    flashplayer: "/local/path/to/jwplayer.flash.swf"
}
```

I can't guarantee that this will work in all situations, but so far it works for me.

**Comment by robwalch** [comment]
@chrisdeely thanks for that info. It's really useful and will others who are running into the same problem.

We have plans to improve player behavior and user experience when Flash is blocked. Part of this requires us being able to detect when this occurs from JavaScript. Another part of the story is for us to do this Flash block checking while preparing to play - rolling it into a new preload feature.

As far as timing, we're working hard to get 7.1 out early next month with some big UI updates for mobile and JavaScript ad capabilities. We'll follow that up quickly with improved support for Flash in Chrome 45+.

**Comment by robwalch** [comment]
Hey @chrisdeely,

We have some changes in v7.1.1 that attempt to deal with the Chrome Power Save feature:

https://github.com/jwplayer/jwplayer/pull/750
https://github.com/jwplayer/jwplayer/pull/751

You may be hosting the swf to get around the issue all together. If not please give this new version a try. It should make the jwplayer UI get out of the way and display Chrome's plugin paused state. This might not be ideal for a small audio player but that's our only option when Chrome throttles Flash and prevents audio and network activity from progressing.

https://ssl.p.jwpcdn.com/player/v/7.1.1/jwplayer.js

In 7.2.x we'll attempt to even better identify this state and add additional text to the player area.

Thanks!

**Comment by chrisdeely** [comment]
Thanks, I'll give it a try. In 7.2, could I suggest a 'setupError' event be thrown so that we could message our users directly?

**Comment by robwalch** [comment]
We will send an error event and include some error messaging in the display in 7.2. It will happen after setup/ready so not a setup error.

We left that out of this patch because we don't have an easy way of identifying the difference between Flash being backgrounded (tab or browser hidden) and Chrome's power saving feature. Both cases send a 'ThrottledEvent' from Flash at which point Flash continues to run but at a slower frame rate. Power save also blocks audio and network access but we don't know of any events that could notify us of this yet. We'll need to observe buffers combined with the throttled state and make an informed assumption.

**Comment by robwalch** [comment]
@chrisdeely,

We've begun beta testing 7.2. If you have some time to give this a try I'd like to know if the Flash throttle and blocking detection helps with your issue. If you're already self hosting the swf it's probably a non-issue but let me know either way.
 
https://ssl.p.jwpcdn.com/player/v/7.2.0-beta.1/jwplayer.js

If your player runs in "audio mode" we still have an open ticket to make the swf that's handling audio small enough to prevent it from being throttled.

**Comment by chrisdeely** [comment]
@robwalch, I do see the throttling detection working, and an error is thrown:
```javascript
{message: "Flash plugin is blocked", type: "error"}
```

A message is properly displayed to the user to inform them of the situation, but there doesn't seem to be anything they can do in response. Clicking on the blocked player doesn't do anything except fire a ```tabFocus``` event.

Perhaps my player area is still too small? I'm testing it at 600x40.

You can see the test code here: http://plnkr.co/edit/PvG5w4?p=preview (just insert a valid JW key in the code)

**Comment by robwalch** [comment]
Yeah this is the situation we're aiming to fix by sizing the object element to 5x5 when the player is too small to make the swf clickable. It didn't make it into this beta, but I'll see that we get it in the release.

**Comment by robwalch** [comment]
@chrisdeely we're releasing v7.2 tomorrow. This is working on our end. When the player is in audio mode, the swf object is scaled to 1x1 px. This prevents Chrome from throttling playback.

Script URL https://ssl.p.jwpcdn.com/player/v/7.2.1/jwplayer.js
